wikipedia_cleanup/evaluation.py

```diff
@@ -0,0 +1,158 @@
+            stats.append(evaluate_prediction(timeframe_label, timeframe_prediction))
+
+    plot_multi_stat_data(stats, testing_timeframes, buckets, output_folder / "bucketed")
+
+
+def evaluate_template_predictions(
+    labels, predictions, testing_timeframes, output_folder, keys, data
+):
+    top_n_templates = 10
+    templates = data.groupby(["template"])
+    template_count = templates["value_valid_from"].count()
+    template_count = template_count.sort_values(ascending=False)
+    top_templates = template_count.head(top_n_templates).index.tolist()
+    keys_of_template = templates["key"].apply(set)
+
+    stats = []
+    for temp in top_templates:
+        current_keys = keys_of_template.loc[temp]
+        used_indices = [key in current_keys for key in keys]
+        cur_labels = [arr[used_indices] for arr in labels]
+        cur_predictions = [arr[used_indices] for arr in predictions]
+        for timeframe_label, timeframe_prediction in zip(cur_labels, cur_predictions):
+            stats.append(evaluate_prediction(timeframe_label, timeframe_prediction))
+
+    plt.figure(figsize=(10, 5))
+    plot_multi_stat_data(
+        stats, testing_timeframes, top_templates, output_folder / "templates"
+    )
+
+
+def plot_multi_stat_data(
+    stats: List[Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]],
+    testing_timeframes: List[int],
+    categories: List[Any],
+    output_file: Path,
+) -> None:
+    assert all(len(pair) == 2 for single_run in stats for pair in single_run)
+    plotting_df = pd.DataFrame(
+        np.array(stats)[..., :2, 1].reshape(-1, 2),
+        columns=["precision", "recall"],
+    )
+    plotting_df["timeframe"] = list(testing_timeframes * len(categories))
+    plotting_df["category"] = list(
+        itertools.chain.from_iterable(
+            ([([i] * len(testing_timeframes)) for i in categories])
+        )
+    )
+    plotting_df = (
+        plotting_df.set_index(["timeframe", "category"])
+        .sort_index()
+        .reset_index()
+        .set_index(["category", "timeframe"])
+    )
+    plotting_df.plot(kind="bar")
+    plt.ylabel("score")
+    plt.savefig(output_file, bbox_inches="tight")
+
+
+def evaluate_metric_over_time(labels, predictions, testing_timeframes, output_folder):
+    for i, timeframe in enumerate(testing_timeframes[:-1]):
+        current_labels = labels[i]
+        current_predictions = predictions[i]
+        stats = [
+            precision_recall_fscore_support(
+                current_labels[:, i],
+                current_predictions[:, i],
+                labels=[1],
+                zero_division=0,
+            )
+            for i in range(current_labels.shape[1])
+        ]
+        prec = np.array([stat[0][0] for stat in stats])
+        rec = np.array([stat[1][0] for stat in stats])
+        plt.figure()
+        if timeframe == 1:
+            average = 5
+            prec = np.mean(np.reshape(prec, (-1, average)), axis=1)
```

Comment:
You assume here that the shape is dividable by 5. That might currently be the case, but might fail later on when choosing a different test set size. Maybe add padding?

Reply:
Solved, i check if its divisible by 5, otherwise i skip it